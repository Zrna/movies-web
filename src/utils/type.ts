/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/ban-ts-comment */
export type WithNestedKeysOf<ObjectType extends object> = {
  [Key in keyof ObjectType & (string | number)]: ObjectType[Key] extends object
    ? // @ts-ignore
      Key | `${Key}.${WithNestedKeysOf<ObjectType[Key]>}`
    : Key;
}[keyof ObjectType & (string | number)];
