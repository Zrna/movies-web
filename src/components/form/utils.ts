import { FieldMetaState } from 'react-final-form';

import { TextInputProps } from '~/ui';

export interface FormFieldProps extends Omit<TextInputProps, 'value' | 'onChange'> {
  name: string;
  type?: string;
  label?: string;
  validate?: any;
  value?: null;
  onChange?(): null;
}

export function getFieldError(meta: FieldMetaState<any>) {
  if (meta.touched && meta.error) {
    return meta.error;
  }

  return meta.submitError;
}
