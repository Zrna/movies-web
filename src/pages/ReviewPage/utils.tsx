import { theme } from '~/ui';

export const reactionsMap: { [key: number]: { icon: theme.Icon; text: string } } = {
  0: { icon: 'reaction0', text: 'No comment' },
  1: { icon: 'reaction1', text: "I wish I hadn't watched" },
  2: { icon: 'reaction2', text: 'Boring' },
  3: { icon: 'reaction3', text: 'Not great, not terrible' },
  4: { icon: 'reaction4', text: "It's great" },
  5: { icon: 'reaction5', text: 'Loving it!' },
};
