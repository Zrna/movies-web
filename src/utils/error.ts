import { AxiosError } from 'axios';

import { showToast } from '~/ui';

export function getErrorMessage(error: AxiosError) {
  if (error.response) {
    return error.response.data.error;
  }

  return 'Something went wrong';
}

export function showErrorToast(error: AxiosError) {
  showToast({
    variant: 'error',
    text: getErrorMessage(error),
  });
}
