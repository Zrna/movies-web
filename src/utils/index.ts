export * from './async';
export * from './error';
export * as validator from './validation';
