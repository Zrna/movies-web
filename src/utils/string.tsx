import { Fragment } from 'react';

export function splitStringToNewLine(str: string) {
  return str.split('\n').map((value, i) => (
    <Fragment key={value + i}>
      {value} <br />
    </Fragment>
  ));
}

export function pluralize(noun: string, length: number) {
  return length === 1 ? noun : `${noun}s`;
}
