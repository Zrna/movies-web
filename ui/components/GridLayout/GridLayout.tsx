import { Property } from 'csstype';
import { Box, BoxProps } from 'theme-ui';

export interface GridLayoutProps extends BoxProps {
  alignItems?: Property.AlignItems | Property.AlignItems[];
  columnGap?: number | number[];
  gap?: number | number[];
  gridTemplateColumns?: Property.GridTemplateColumns | Property.GridTemplateColumns[];
  gridTemplateRows?: Property.GridTemplateRows | Property.GridTemplateRows[];
  justifyItems?: Property.JustifyItems | Property.JustifyItems[];
  rowGap?: number | number[];
}

export const GridLayout: React.FC<GridLayoutProps> = ({
  alignItems,
  columnGap,
  gap,
  gridTemplateColumns,
  gridTemplateRows,
  justifyItems,
  rowGap,
  sx = {},
  ...rest
}) => {
  return (
    <Box
      sx={{
        ...sx,
        alignItems,
        columnGap,
        display: 'grid',
        gap,
        gridTemplateColumns,
        gridTemplateRows,
        justifyItems,
        rowGap,
      }}
      {...rest}
    />
  );
};
