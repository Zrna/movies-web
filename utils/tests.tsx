import { cleanup, fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'theme-ui';

import { theme } from '~/ui';

export { cleanup, fireEvent, render, userEvent };

export function renderWithTheme(Component: JSX.Element) {
  return render(<ThemeProvider theme={theme}>{Component}</ThemeProvider>);
}

export function getThemeColorString(color: theme.Color) {
  return `var(--theme-ui-colors-${color})`;
}

export function expectTextVariant(element: HTMLElement, variant: theme.TextVariant) {
  const { fontFamily, fontSize, fontStyle, fontWeight, lineHeight } = theme.text[variant];

  expect(element).toHaveStyleRule('font-family', fontFamily);
  expect(element).toHaveStyleRule('font-style', fontStyle);
  expect(element).toHaveStyleRule('font-size', fontSize);
  expect(element).toHaveStyleRule('font-weight', fontWeight.toString());
  expect(element).toHaveStyleRule('line-height', lineHeight);

  if (variant === 'label-uppercase') {
    expect(element).toHaveStyleRule('text-transform', theme.text[variant].textTransform);
  }
}

export function mockWindowMatchMedia() {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
}
