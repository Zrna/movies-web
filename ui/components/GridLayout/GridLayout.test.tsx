import { screen, within } from '@testing-library/react';

import { mockWindowMatchMedia, renderWithTheme } from '~/utils/tests';

import { GridLayout } from './GridLayout';

describe('GridLayout', () => {
  beforeAll(mockWindowMatchMedia);

  it('should render', () => {
    renderWithTheme(
      <GridLayout alignItems="center" data-testid="grid" gap={4} gridTemplateColumns="1fr">
        <div>Item 1</div>
        <div>Item 2</div>
        <div>Item 3</div>
      </GridLayout>,
    );

    expect(screen.getByTestId('grid')).toMatchSnapshot();
  });

  it('should render grid component', () => {
    renderWithTheme(<GridLayout data-testid="grid" />);

    expect(screen.getByTestId('grid')).toHaveStyle({ display: 'grid' });
  });

  describe('grid container', () => {
    it('should apply proper columns', () => {
      renderWithTheme(<GridLayout data-testid="grid1" gridTemplateColumns="none" />);
      renderWithTheme(<GridLayout data-testid="grid2" gridTemplateColumns="1fr" />);
      renderWithTheme(<GridLayout data-testid="grid3" gridTemplateColumns="100px 1fr" />);
      renderWithTheme(<GridLayout data-testid="grid4" gridTemplateColumns="repeat(3, 200px)" />);

      expect(screen.getByTestId('grid1')).toHaveStyle({ gridTemplateColumns: 'none' });
      expect(screen.getByTestId('grid2')).toHaveStyle({ gridTemplateColumns: '1fr' });
      expect(screen.getByTestId('grid3')).toHaveStyle({ gridTemplateColumns: '100px 1fr' });
      expect(screen.getByTestId('grid4')).toHaveStyle({ gridTemplateColumns: 'repeat(3, 200px)' });
    });

    it('should align items properly', () => {
      renderWithTheme(<GridLayout alignItems="center" data-testid="grid1" />);
      renderWithTheme(<GridLayout alignItems="start" data-testid="grid2" />);
      renderWithTheme(<GridLayout alignItems="end" data-testid="grid3" />);
      renderWithTheme(<GridLayout alignItems="stretch" data-testid="grid4" />);
      renderWithTheme(<GridLayout alignItems="baseline" data-testid="grid5" />);

      expect(screen.getByTestId('grid1')).toHaveStyle({ alignItems: 'center' });
      expect(screen.getByTestId('grid2')).toHaveStyle({ alignItems: 'start' });
      expect(screen.getByTestId('grid3')).toHaveStyle({ alignItems: 'end' });
      expect(screen.getByTestId('grid4')).toHaveStyle({ alignItems: 'stretch' });
      expect(screen.getByTestId('grid5')).toHaveStyle({ alignItems: 'baseline' });
    });

    it('should justify items properly', () => {
      renderWithTheme(<GridLayout data-testid="grid1" justifyItems="center" />);
      renderWithTheme(<GridLayout data-testid="grid2" justifyItems="start" />);
      renderWithTheme(<GridLayout data-testid="grid3" justifyItems="end" />);

      expect(screen.getByTestId('grid1')).toHaveStyle({ justifyItems: 'center' });
      expect(screen.getByTestId('grid2')).toHaveStyle({ justifyItems: 'start' });
      expect(screen.getByTestId('grid3')).toHaveStyle({ justifyItems: 'end' });
    });
  });

  describe('child items', () => {
    it('should render child items', () => {
      renderWithTheme(
        <GridLayout data-testid="grid">
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
        </GridLayout>,
      );

      const gridElement = screen.getByTestId('grid');

      expect(within(gridElement).getByText('Item 1')).toBeInTheDocument();
      expect(within(gridElement).getByText('Item 2')).toBeInTheDocument();
      expect(within(gridElement).getByText('Item 3')).toBeInTheDocument();
    });
  });

  describe('spacing', () => {
    it('should apply the gap property', () => {
      renderWithTheme(
        <GridLayout data-testid="grid" gap={4}>
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
        </GridLayout>,
      );

      const gridElement = screen.getByTestId('grid');

      expect(gridElement).toHaveStyle({ gap: '16px' });
    });
  });
});
