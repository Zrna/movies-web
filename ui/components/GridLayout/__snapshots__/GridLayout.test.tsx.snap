// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridLayout should render 1`] = `
<div
  class="css-1rvnl0y-Box"
  data-testid="grid"
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
  <div>
    Item 3
  </div>
</div>
`;
